include/fiction/io/read_gate_level_layout.hpp

```diff
@@ -0,0 +1,349 @@
+//
+// Created by simon on 25.09.23.
+//
+
+#ifndef FICTION_READ_GATE_LEVEL_LAYOUT_HPP
+#define FICTION_READ_GATE_LEVEL_LAYOUT_HPP
+
+#include "fiction/traits.hpp"
+#include "fiction/utils/name_utils.hpp"
+
+#include <fiction/layouts/cartesian_layout.hpp>
+#include <fiction/layouts/clocked_layout.hpp>
+#include <fiction/layouts/gate_level_layout.hpp>
+#include <fiction/layouts/tile_based_layout.hpp>
+
+#include <kitty/constructors.hpp>
+#include <tinyxml2.h>
+
+#include <algorithm>
+#include <cctype>
+#include <exception>
+#include <fstream>
+#include <istream>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+namespace fiction
+{
+
+/**
+ * Exception thrown when an error occurs during parsing of an gate_level file.
+ */
+class gate_level_parsing_error : public std::runtime_error
+{
+  public:
+    explicit gate_level_parsing_error(const std::string_view& msg) noexcept : std::runtime_error(msg.data()) {}
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class read_gate_level_layout_impl
+{
+  public:
+    read_gate_level_layout_impl(std::istream& s, const std::string_view& name) : lyt{}, is{s}
+    {
+        set_name(lyt, name);
+    }
+
+    read_gate_level_layout_impl(Lyt& tgt, std::istream& s) : lyt{tgt}, is{s} {}
+
+    Lyt run()
+    {
+        // tinyXML2 does not support std::istream, so we have to read the whole file into a string first
+        std::stringstream buffer{};
+        buffer << is.rdbuf();
+        const std::string sqd_content{buffer.str()};
+
+        tinyxml2::XMLDocument xml_document{};
+        xml_document.Parse(sqd_content.c_str());
+
+        if (xml_document.ErrorID() != 0)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: " + std::string(xml_document.ErrorName()));
+        }
+
+        auto* const fcn_root = xml_document.FirstChildElement("fcn");
+        if (fcn_root == nullptr)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: no root element 'fcn'");
+        }
+
+        auto* const layout = fcn_root->FirstChildElement("layout");
+        if (layout == nullptr)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: no element 'layout'");
+        }
+
+        auto* const name = layout->FirstChildElement("name");
+        if (name != nullptr and name->GetText())
+        {
+            std::string layout_name = name->GetText();
+            set_name(lyt, layout_name);
+        }
+
+        auto* const clocking = layout->FirstChildElement("clocking");
+        if (clocking != nullptr)
+        {
+            lyt.replace_clocking_scheme(*get_clocking_scheme<Lyt>(clocking->GetText()));
```

Comment:
Btw: Custom schemes are possible via override of the clock zones. Is that feature already possible in your file format version?